lib/spack/docs/testing_guide.rst

```diff
@@ -0,0 +1,299 @@
+            <ElapsedMinutes>0</ElapsedMinutes>
+        </Build>
+    </Site>
+
+Spack uses a short version of each package's spec as the CDash build
+name, which will be shown in the dashboard.
+
+For more detailed output, you can use the ``cdash-complete`` format:
+
+.. code-block:: console
+
+   $ spack install --log-file report --log-format=cdash-complete libelf
+
+This will create separate files: ``report.build.xml``,
+``report.configure.xml``, and ``report.test.xml``, for the build,
+configure, and tests steps, respectively.
+
+If you want to upload these fils to a CDash instance, you can use ``curl``:
+
+.. code-block:: console
+
+   $ curl --upload-file report.build.xml https://example.com/cdash/submit.php?project=<projectname>
+   $ curl --upload-file report.configure.xml https://example.com/cdash/submit.php?project=<projectname>
+   $ curl --upload-file report.test.xml https://example.com/cdash/submit.php?project=<projectname>
+
+Spack can also automate this step for you as part of the ``spack
+test-suite`` command described in the next section.
+
+
+.. _cmd-spack-test-suite:
+
+---------------------
+CDash test suites
+---------------------
+
+The ``spack test-suite`` command reads in a specialy formatted YAML file
+describing a set of combinatorial tests.  It can be used to easily test a
+package or suite of packages with many different compilers and build
+options.
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Test suite YAML format
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Here's an example file:
+
+.. code-block:: yaml
+
+   #
+   # This YAML file describes a Spack test suite
+   #
+   test-suite:
+       #
+       # Optional include/exclude spec lists and upload information.
+       #
+
+       # Only specs that match a spec in this list will be included in
+       # the tests.  If include is missing, all specs are built.
+       include: [bzip2, libelf, libdwarf]
+
+       # Specs that match a spec in this list are excluded.
+       # If exclude is missing or empty, all included packages are built.
+       exclude: []
+
+       # URL of the cdash server where results should be submitted.
+       # Optional. Defaults to https://spack.io/cdash
+       - cdash: ["https://spack.io/cdash"]
+
+       # Project on the cdash server where results should be submitted.
+       # Optional. Defaults to 'spack'.
+       - project: spack
+
+       # Build matrix.  Spack takes the cartesian product of each dimension
+       # in this section to define a set of specs to build.
+       matrix:
+         # List of packages, each with a set of versions to test.
+         - packages:
+             abinit:
+               versions: [8.0.8b]
+             ack:
+               versions: [2.14]
+
+         # List of compiler versions. Each package is tested with all
+         # compiler versions.
+         - compilers:
+             gcc:
+               versions: [4.9.0, 4.7.1, 4.6.3, 4.6.1]
+             clang:
+               versions: [7.3.0, 3.4, 3.1]
```

Comment:
That's the intent with `spack.util.spec_set`.  The way I've reworked it, you can basically cartesian-product arbitrary lists of specs.  The `compilers` and `packages` elements are shorthand for particular types of spec lists.

The schema is currently written with `spack test-suite` in mind, but it can be factored out to 
 a `suite.yaml` or maybe `stack.yaml`.  Which do you like better?  I kind of like `stack.yaml` and thought about calling the format `spack-stack` instead of `test-suite`.